Reference for Issue #210

a-y-a-n-das opened the pull request:
# Pull Request

## Description
Partial implementation of automated error monitoring for Vercel deployments.

Initially, I was fetching logs from a sample file for testing.
When running against a Vercel deployment, it became clear that Vercel only provides runtime logs, which are ephemeral.
As a result, continuous monitoring would be required to catch errors in real-time.
This PR includes the partial script for reference and as a starting point for future implementation.

## Related Issues

Relates to #210

## Changes Made

<!-- List the specific changes made in this PR -->

- [x] Added scripts/monitor-vercel-errors.js partial script.
- [x] Parses multi-line error blocks and deduplicates using SHA1 hashes stored in error-hashes.json.
- [x] Demonstrates GitHub issue creation using the gh CLI for new error blocks.

## Type of Change

<!-- Mark the relevant option with an "x" -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactoring (code improvement without changing functionality)
- [ ] Documentation update
- [ ] Performance improvement
- [ ] Test coverage improvement
- [ ] Dependency update

## Testing

<!-- Describe the testing you've done -->

- [ ] Unit tests added/updated
- [ ] Integration tests added/updated
- [ ] E2E tests added/updated
- [ ] Manual testing completed

**Test Details:**

<!-- Describe your test cases and results -->

## Quality Checklist

<!-- Ensure all items are checked before submitting -->

- [x] All tests passing locally (`pnpm test`)
- [x] No ESLint errors (`pnpm lint`)
- [x] Code formatted with Prettier (`pnpm prettier`)
- [x] All files ≤150 lines (verify with: `find . -name "*.ts" -o -name "*.tsx" -o -name "*.js" | xargs wc -l | awk '$1 > 150'`)
- [x] Test coverage maintained or improved
- [x] No new security vulnerabilities (`pnpm audit`)
- [ ] Documentation updated (if needed)
- [x] TypeScript types are proper (no `any` unless necessary)

## Performance Impact

<!-- Describe any performance implications of your changes -->

- [x] No performance impact
- [ ] Performance improved
- [ ] Potential performance impact (explain below)

**Details:**

## Breaking Changes

<!-- List any breaking changes and migration steps -->

- [x] No breaking changes
- [ ] Breaking changes (describe below)

**Details:**

## Screenshots/Demos

<!-- If UI changes, include screenshots or GIFs -->

## Additional Notes

<!-- Any additional information reviewers should know -->

---
Additional Notes

Vercel runtime logs are ephemeral; historical logs are not available.
Full continuous monitoring and real-time error detection will require additional implementation.

**For Reviewers:**

- Does this PR follow the [contribution guidelines](../CONTRIBUTING.md)?
- Are all quality checks passing?
- Is the code well-tested and documented?
- Are there any security concerns?


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- New Features
  - Added server-side error tracking via Sentry for improved observability and error reporting.
  - Added a test API endpoint to validate and send errors to Sentry.

- Chores
  - Added an automated workflow to fetch recent Sentry issues and create GitHub issues on schedule or on demand.
  - Installed Sentry runtime dependency and exposed SENTRY_DSN to the build environment.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

a-y-a-n-das commented on scripts/monitor-vercel-errors.js:
```diff
@@ -0,0 +1,79 @@
+    }
+  });
+
+  fs.writeFileSync(hashFile, JSON.stringify(existingHashes, null, 2));
+  if (newErrors.length === 0) console.log('No new unique errors to report.');
```
yup, Let someone else do it... ;)

thomasdavis commented:
## Code Review - Partial Implementation Analysis

Thank you for the contribution! This is a great start on automated error monitoring. However, the current implementation has several issues that need resolution before it can be merged.

### Critical Issues ❌

1. **Hardcoded Personal Repository** (Line 62)
   ```js
   gh issue create --repo a-y-a-n-das/jsonresume.org
   ```
   This creates issues in your fork, not the main repo. Should be:
   ```js
   --repo jsonresume/jsonresume.org
   ```

2. **Hardcoded Deployment URL** (Line 5)
   ```js
   const DEPLOYMENT_URL = "https://jsonresume-org-homepage2-9zdb0ytck-ayan-das-projects-afcc9401.vercel.app";
   ```
   This is your personal deployment. Should use environment variable from workflow.

3. **Security Risk: Command Injection** (Lines 62-64)
   The `title` and `body` are directly interpolated into shell command without escaping. This could break or expose vulnerabilities if error messages contain quotes or special chars.

### Architecture Issues ⚠️

4. **Vercel Log Limitation**
   As noted in your PR description, Vercel only provides ephemeral runtime logs. This means:
   - Historical errors are lost
   - Script only sees errors during the exact moment it runs
   - Most runs will find zero errors (as you discovered)

5. **Error Detection Logic** (Lines 38-47)
   - Assumes errors are followed by blank lines (fragile)
   - Misses multi-line errors that don't end with blank line
   - Could create false positives from log messages containing "error" keyword

### Recommended Fixes

```js
// Fix hardcoded values - use env vars from workflow
const DEPLOYMENT_URL = process.env.VERCEL_DEPLOYMENT_URL;
const REPO = process.env.GITHUB_REPOSITORY || 'jsonresume/jsonresume.org';

// Fix command injection - write to temp file instead
const tempFile = `/tmp/issue-body-${Date.now()}.md`;
fs.writeFileSync(tempFile, body);
execSync(`gh issue create --repo ${REPO} --title "${title.replace(/"/g, '\\"')}" --body-file ${tempFile} --label "vercel-error"`, { stdio: 'inherit' });
fs.unlinkSync(tempFile);
```

### Alternative Approach Recommendation

Given Vercel's ephemeral logging limitation, consider:

1. **Sentry/LogRocket Integration** - Real-time error tracking with full stack traces
2. **Vercel Integration API** - Use Vercel's integration events for error notifications
3. **Application-Level Logging** - Add error boundaries that POST to a logging endpoint
4. **Edge Function Monitoring** - Monitor edge function errors via Vercel's dashboard API

The current batch approach will miss 99% of errors since logs are ephemeral.

### Next Steps

Would you like to:
- A) Fix the issues above and keep the batch monitoring approach (limited effectiveness)
- B) Pivot to real-time error monitoring (Sentry/LogRocket/etc.)
- C) Close this PR as reference and create a new issue for improved error monitoring design

Let me know your preference and I'm happy to help guide the implementation!

a-y-a-n-das commented:
Thank you for the detailed suggestions! I like the approach you recommended. Although I’m still learning, I can try to implement the Real-time error logic using Sentry/LogRocket.

a-y-a-n-das commented:
@thomasdavis 

- Added configuration for Sentry to capture runtime and build-time errors automatically and send them to the Sentry dashboard.
- Set up environment variables and ensured they’re read dynamically (no hardcoding).
- Configured the GitHub Action to fetch and process Sentry issues into GitHub.
- Added a test-error.js endpoint to simulate runtime errors for validation.

Note: Please update the current error-catching lines to use Sentry’s error handling methods.

a-y-a-n-das commented on scripts/sentry-to-github.js:
```diff
@@ -0,0 +1,69 @@
+  return res.json();
+}
+
+// Check if GitHub issue exists by title
+async function githubIssueExists(title) {
+  const url = `https://api.github.com/repos/${GITHUB_REPO}/issues?state=open&per_page=100`;
+  const res = await fetch(url, {
+    headers: { Authorization: `token ${GITHUB_TOKEN}` },
+  });
+  const issues = await res.json();
+  return issues.some((issue) => issue.title === title);
+}
```
Yup, it can be implemented later

a-y-a-n-das commented on .github/workflows/fetch-issue.yml:
```diff
@@ -0,0 +1,67 @@
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          MY_GITHUB_REPO: ${{ secrets.MY_GITHUB_REPO }}
+          SENTRY_DSN: ${{ secrets.SENTRY_DSN }}
+
```
it has been handled inside the script.

a-y-a-n-das commented on scripts/sentry-to-github.js:
```diff
@@ -0,0 +1,69 @@
+const SENTRY_ORG = process.env.SENTRY_ORG;
+const SENTRY_PROJECT = process.env.SENTRY_PROJECT;
+const GITHUB_TOKEN = process.env.GITHUB_TOKEN;
+const GITHUB_REPO = process.env.GITHUB_REPO; // "owner/repo" format
```
can be done

a-y-a-n-das commented:
@coderabbit re-review
